qiskit_optimization/applications/SK_model.py

```diff
@@ -0,0 +1,104 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+from typing import Optional, Union, List
+
+import networkx as nx
+import numpy as np
+from docplex.mp.model import Model
+
+from .graph_optimization_application import OptimizationApplication
+from ..algorithms import OptimizationResult
+from ..problems.quadratic_program import QuadraticProgram
+from ..translators import from_docplex_mp
+
+
+class SKModel(OptimizationApplication):
+    """Optimization application for the "Sherrington Kirkpatrick (SK) model" [1].
+
+    The SK Hamiltonian over n spins is given as:
+        H(x)=-1/\sqrt{n} \sum_{i<j} w_{i,j}x_ix_j, where x_i\in\pm 1
+    and w_{i,j}\in\pm 1 are chosen independently and uniformly at random.
+    Notice that there are other variants e.g., with w_{i,j} chosen from the normal distribution
+    with mean 0 and variance 1.
+
+    References:
+        [1]: Dmitry Panchenko. "The Sherrington-Kirkpatrick model: an overview",
+        https://arxiv.org/abs/1211.1094
+    """
+
+    def __init__(self, n: int, rng: Optional[np.random.RandomState] = np.random):
+        """
+        Constructor for the SK model.
+
+        Args:
+            n: number of sites
+            rng: numpy pseudo-random number generator
+        """
+        self._rng = rng
+        self._n = n
+        self._graph = nx.complete_graph(self._n)
+
+        self.new_instance()
+
+    def new_instance(self) -> None:
+        """Generate a new instance of the SK model"""
+        for _, _, d in self._graph.edges(data=True):
+            d['weight'] = self._rng.choice([-1, 1])
+
+    def to_quadratic_program(self) -> QuadraticProgram:
+        """Convert an SK model problem instance into a
+        :class:`~qiskit_optimization.problems.QuadraticProgram`
+
+        Returns:
+            The :class:`~qiskit_optimization.problems.QuadraticProgram` created
+            from the SK problem instance.
+        """
+        mdl = Model(name="SK-model")
+        x = {
```

Comment:
Can we think of a more descriptive name?

Reply:
I took it from MaxCut. There is `name="Max-cut"`.

Reply:
Sorry, I had `x` in mind.

Reply:
It is a vector of binary variables that is called `x`. Do you think it requires a different name?

Reply:
Maybe `decision_vars` or sth along these lines?

Reply:
I have couple reasons why `x` is better:
1) it makes the objective function more readable.
2) it seems to be a convention, please see other applications max_cut, knapsack, graph_partition, ... if you want to change it it should be refactored in the whole module.
3) I like `x`. :)) ok, it is better aligned to what you see in the literature then `decision_vars`. Also it is an internal variable in this function.